server/src/main/java/org/elasticsearch/common/util/BigArrays.java

```diff
@@ -773,22 +691,32 @@ public DoubleArray grow(DoubleArray array, long minSize) {
         return resize(array, newSize);
     }
 
-    public static class DoubleBinarySearcher extends BinarySearcher<Double>{
+    public static class DoubleBinarySearcher extends BinarySearcher{
 
         DoubleArray array;
+        double searchFor;
 
         public DoubleBinarySearcher(DoubleArray array){
             this.array = array;
+            this.searchFor = Integer.MIN_VALUE;
         }
 
         @Override
-        protected int compare(int index, Double value) {
-            return Double.compare(array.get(index), value);
+        protected int compare(int index) {
+            // Prevent use of BinarySearcher.search() and force the use of DoubleBinarySearcher.search()
+            assert this.searchFor != Integer.MIN_VALUE;
+
+            return Double.compare(array.get(index), searchFor);
         }
 
         @Override
-        protected double distance(int index, Double value) {
-            return Math.abs(array.get(index) - value);
+        protected double distance(int index) {
+            return Math.abs(array.get(index) - searchFor);
+        }
+
+        public int search(int from, int to, double searchFor) {
+            this.searchFor = searchFor;
+            return super.search(from, to);
```

Comment:
Maybe `search` should be `protected` if it only makes sense to call it from a subclass? Then you wouldn't need the assert.